File: graphs/max_area_of_island.py
# Q:695: https://leetcode.com/problems/max-area-of-island/

from typing import List

class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        visited, maximumArea = {}, 0

        def dfs(row, col):
            if row < 0 or row >= len(grid) or col < 0 or col >= len(grid[0]) or grid[row][col] == 0 or (row, col) in visited:
                return 0

            visited[(row, col)] = True

            directions = [[0, 1], [0, -1], [-1, 0], [1, 0]]

            total = 0

            for dir in directions:
                newRow, newCol = row + dir[0], col + dir[1]

                total += dfs(newRow, newCol)
            
            return 1 + total


        for idx, _ in enumerate(grid):
            for colIdx, _col in enumerate(grid[0]):
                currentMax = dfs(idx, colIdx)
                maximumArea = max(currentMax, maximumArea)

        return maximumArea


c = Solution()


grid1 = [[0,0,1,0,0,0,0,1,0,0,0,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,1,1,0,1,0,0,0,0,0,0,0,0],[0,1,0,0,1,1,0,0,1,0,1,0,0],[0,1,0,0,1,1,0,0,1,1,1,0,0],[0,0,0,0,0,0,0,0,0,0,1,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,0,0,0,0,0,0,1,1,0,0,0,0]]


grid2 = [[0,0,0,0,0,0,0,0]]

print(c.maxAreaOfIsland(grid1))

File: graphs/max_area_of_island.js
// Q: 695 - https://leetcode.com/problems/max-area-of-island/submissions/

/**
 * @param {number[][]} grid
 * @return {number}
 */
 var maxAreaOfIsland = function(grid) {
    const visited = new Set();
    let maxLength = 0;
    
    const dfs = (row, col) => {
        if (
            row < 0 ||
            row >= grid.length ||
            col < 0 ||
            col >= grid[0].length ||
            visited.has(`${row},${col}`) ||
            grid[row][col] === 0
        ) { 
            return 0;
        }
        
        visited.add(`${row},${col}`);
        
        let count = 1;
                
        const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];
        
        for(let dir of directions){
            const [newRow, newCol] = [dir[0] + row, dir[1] + col];
                        
            count += dfs(newRow, newCol)
        }
        
        return count;
    }
    
    for(let i = 0; i < grid.length; i++){
        for(let j = 0; j < grid[0].length; j++){
            if(grid[i][j] === 1 && !visited.has(`${i},${j}`)){
                const area = dfs(i, j, 0)
                if (area > maxLength){
                    maxLength = area;
                }
            }
        }
    }
    
    return maxLength;
};